trin-core/src/portalnet/overlay_service.rs

```diff
@@ -530,14 +627,39 @@ where
         .await
     }
 
-    /// Returns an ENR if one is known for the given NodeId in overlay routing table
-    pub fn find_enr(&self, node_id: &NodeId) -> Option<Enr> {
-        // check if we know this node id in our routing table
-        let key = kbucket::Key::from(*node_id);
-        if let kbucket::Entry::Present(entry, _) = self.kbuckets.write().entry(&key) {
-            return Some(entry.value().enr.clone());
-        }
-        None
+    /// Maintains the query pool.
+    ///
+    /// Returns a `FindNodeQueryEvent` when the `QueryPoolState` updates.
+    /// This happens when a query needs to send a request to a node, when a query has completed,
+    // or when a query has timed out.
+    async fn query_event_poll(
+        queries: &mut QueryPool<NodeId, Vec<NodeId>, Vec<NodeId>, FindNodeQuery<NodeId>>,
+    ) -> FindNodeQueryEvent {
+        future::poll_fn(move |_cx| match queries.poll() {
+            QueryPoolState::Finished(query_id, query) => {
+                Poll::Ready(FindNodeQueryEvent::Finished(query_id, query))
+            }
+            QueryPoolState::Timeout(query_id, query) => {
+                warn!("Query id: {:?} timed out", query_id);
+                Poll::Ready(FindNodeQueryEvent::TimedOut(query_id, query))
+            }
+            QueryPoolState::Waiting(Some((query_id, query_info, query, return_peer))) => {
+                let node_id = return_peer;
+
+                let request_body = match query_info.rpc_request(return_peer) {
+                    Ok(request_body) => request_body,
+                    Err(_) => {
+                        query.on_failure(&node_id);
+                        return Poll::Pending;
+                    }
+                };
```

Comment:
Hmm, I thought about it and I think it's fine as-is. What do you think it should do in that case instead?

Reply:
IIUC the failure should not adversely affect the query's progress. We should be able to prevent that match arm from occurring though. In `advance_query`, we do not filter out an ENR that is equal to that of the local node. We could add that to prevent this from occurring. This filtering does occur in `process_discovered_nodes`, and it may be preferable to move the `advance_query` logic into that function. That is, as long as we could apply the same function for ENRs returned in a `Content` message that may belong to a `FindContentQuery`.

Reply:
I added a filter in `advance_query` to make sure that this case doesn't happen. Though I kept the `Result` return type from `rpc_request` because we still need a way to denote the idea that log2 of 0 is undefined. I tried using an `Option<Request>` return instead and it wasn't any cleaner imo. 

And yea, I think that `advance_query` needs to be done outside of `process_discovered_enrs` because I think that a different function (`advance_findnode_query`/`advance_content_query`) will need to be called for advancing a `FindContent` query.